mautic/mautic
# Improved contact frequency rules with enforcement option and dynamic time spans

| Q  | A
| --- | ---
| Bug fix? | No
| New feature? | Yes
| Automated tests included? | Yes (this needs confirmation)
| BC breaks? | None
| Deprecations? | None

#### Description:
This update is primarily about **frequency rules** in Mautic. Improves multiple things for system-wide and lead-specific contact frequency rules.

**Affected bundles:**
- CoreBundle
- EmailBundle
- LeadBundle
- ChannelBundle
- ConfigBundle
- SmsBundle

**DB migrations:**
- 20181215173251 – Adds a new column `frequency_unit` to table `lead_frequencyrules`.

**Changes:**
- Provides **a new type of time unit, `minute`**, for more miniscule rate limiting in API sending
- Currently in Mautic code and panels, you can only have rules fixed with "1" repetition, like "1 minute", "1 week" etc. This PR provides `frequency_units` for Lead frequency rules, thus making rules like "2 minutes", "3 hours" possible.
- Provides a new configuration `enforce_frequency_check` to apply frequency rule checks for transactional emails as well. They are currently only applied on marketing emails. I don't really like the way this variable behaves actually. I'd rather have an array: `["transactional", "marketing", "push"]` etc.
- All 3 changes have been reflected on Mautic dashboards. Affected areas: `Contact Detail > Preferences` and `Configuration > Email Settings`

**Warning:**
Emails sent manually on Mautic dashboards are not applied frequency rules. So I kept this behavior.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Go to Configuration > Email settings. Confirm it has an input for unit amount and reads like "each X Y". When saved, should go to your local.php
3. Go to a Contact. On detail menu on top-right, click Preferences. You should be able to create and save rules like "5 times in 3 minutes".

+ if we already have a test plan for testing email sending, scheduling etc., they should probably be done for this PR. For integration, I did tests over the API.

**Is anything missing?** This is my first PR here, and first days with Symfony framework. Please be aware of that and tell me any changes you may need :)
